File: google-cloud-bigquery-storage/src/test/scala/akka/stream/alpakka/googlecloud/bigquery/storage/scaladsl/BigQueryAvroStorageSpec.scala
/*
 * Copyright (C) since 2016 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.stream.alpakka.googlecloud.bigquery.storage.scaladsl

import akka.stream.alpakka.googlecloud.bigquery.storage.impl.AvroDecoder
import akka.stream.alpakka.googlecloud.bigquery.storage.{
  BigQueryRecord,
  BigQueryStorageSettings,
  BigQueryStorageSpecBase
}
import akka.stream.alpakka.testkit.scaladsl.LogCapturing
import akka.stream.scaladsl.Sink
import org.scalatest.BeforeAndAfterAll
import org.scalatest.wordspec.AnyWordSpecLike
import org.scalatest.matchers.should.Matchers

class BigQueryAvroStorageSpec
    extends BigQueryStorageSpecBase(21002)
    with AnyWordSpecLike
    with BeforeAndAfterAll
    with Matchers
    with LogCapturing {

  "BigQueryAvroStorage.readAvro" should {
    val avroSchema = storageAvroSchema.value
    val avroRows = storageAvroRows.value

    "stream the results for a query in records merged" in {
      val decoder = AvroDecoder(avroSchema.schema)
      val records = decoder.decodeRows(avroRows.serializedBinaryRows).map(gr => BigQueryRecord.fromAvro(gr))

      BigQueryAvroStorage
        .readRecordsMerged(Project, Dataset, Table, None)
        .withAttributes(mockBQReader())
        .runWith(Sink.seq)
        .futureValue shouldBe Seq.fill(DefaultNumStreams * ResponsesPerStream)(records)
    }

    "stream the results for a query in records" in {
      val decoder = AvroDecoder(avroSchema.schema)
      val records = decoder.decodeRows(avroRows.serializedBinaryRows).map(gr => BigQueryRecord.fromAvro(gr))

      BigQueryAvroStorage
        .readRecords(Project, Dataset, Table, None)
        .withAttributes(mockBQReader())
        .map(a => a.reduce((a, b) => a.merge(b)))
        .flatMapMerge(100, identity)
        .runWith(Sink.seq)
        .futureValue shouldBe Seq.fill(DefaultNumStreams * ResponsesPerStream)(records).flatten
    }

    "stream the results for a query merged" in {
      BigQueryAvroStorage
        .readMerged(Project, Dataset, Table, None)
        .withAttributes(mockBQReader())
        .map(s => s._2.map(b => (s._1, b)))
        .flatMapMerge(100, identity)
        .runWith(Sink.seq)
        .futureValue shouldBe Vector.fill(DefaultNumStreams * ResponsesPerStream)((avroSchema, avroRows))
    }

    "stream the results for a query" in {
      BigQueryAvroStorage
        .read(Project, Dataset, Table, None)
        .withAttributes(mockBQReader())
        .map { s =>
          s._2
            .reduce((a, b) => a.merge(b))
            .map(b => (s._1, b))
        }
        .flatMapMerge(100, identity)
        .runWith(Sink.seq)
        .futureValue shouldBe Vector.fill(DefaultNumStreams * ResponsesPerStream)((avroSchema, avroRows))
    }
  }

  def mockBQReader(host: String = bqHost, port: Int = bqPort) = {
    val reader = GrpcBigQueryStorageReader(BigQueryStorageSettings(host, port))
    BigQueryStorageAttributes.reader(reader)
  }

  override def beforeAll(): Unit = {
    super.beforeAll()
    startMock()
  }

  override def afterAll(): Unit = {
    stopMock()
    system.terminate()
    super.afterAll()
  }

}

File: google-cloud-bigquery-storage/src/test/java/akka/stream/alpakka/googlecloud/bigquery/storage/javadsl/BigQueryAvroStorageSpec.java
/*
 * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.stream.alpakka.googlecloud.bigquery.storage.javadsl;

import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;

import org.apache.avro.Schema;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import com.google.cloud.bigquery.storage.v1.avro.AvroRows;
import com.google.cloud.bigquery.storage.v1.avro.AvroSchema;

import akka.stream.Attributes;
import akka.stream.Materializer;
import akka.stream.alpakka.googlecloud.bigquery.storage.BigQueryRecord;
import akka.stream.alpakka.googlecloud.bigquery.storage.BigQueryStorageSettings;
import akka.stream.alpakka.googlecloud.bigquery.storage.BigQueryStorageSpecBase;
import akka.stream.alpakka.testkit.javadsl.LogCapturingJunit4;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.util.ByteString;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.stream.Collectors;

import static org.junit.Assert.assertEquals;
import scala.compat.java8.FutureConverters.FutureOps;

public class BigQueryAvroStorageSpec extends BigQueryStorageSpecBase {

  private final AvroSchema avroSchema = storageAvroSchema().value();
  private final AvroRows avroRows = storageAvroRows().value();

  private final Schema schema = new Schema.Parser().parse(avroSchema.schema());
  private final AvroByteStringDecoder avroByteStringDecoder = new AvroByteStringDecoder(schema);

  @Rule public final LogCapturingJunit4 logCapturing = new LogCapturingJunit4();

  private List<BigQueryRecord> expectedRecords;

  @Before
  public void setUp() throws Exception {
    ByteString byteString = ByteString.fromArray(avroRows.serializedBinaryRows().toByteArray());
    expectedRecords =
        new FutureOps<>(
                avroByteStringDecoder.apply(
                    byteString, system().dispatcher(), Materializer.matFromSystem(system())))
            .toJava()
            .toCompletableFuture()
            .get();
  }

  @Test
  public void shouldReturnResultsForAQueryInMergedRecords()
      throws ExecutionException, InterruptedException, TimeoutException {
    List<BigQueryRecord> bigQueryRecords =
        BigQueryAvroStorage.readRecordsMerged(Project(), Dataset(), Table())
            .withAttributes(mockBQReader()).runWith(Sink.seq(), system()).toCompletableFuture()
            .get(5, TimeUnit.SECONDS).stream()
            .flatMap(Collection::stream)
            .collect(Collectors.toList());

    assertEquals("List of records should be the same", expectedRecords, bigQueryRecords);
  }

  @Test
  public void shouldReturnResultsForAQueryInRecords()
      throws ExecutionException, InterruptedException, TimeoutException {
    List<BigQueryRecord> bigQueryRecords =
        BigQueryAvroStorage.readRecords(Project(), Dataset(), Table())
            .withAttributes(mockBQReader())
            .map(a -> a.stream().reduce(Source::merge))
            .filter(Optional::isPresent)
            .map(Optional::get)
            .flatMapConcat(a -> a)
            .runWith(Sink.seq(), system())
            .toCompletableFuture()
            .get(5, TimeUnit.SECONDS);

    assertEquals("List of records should be the same", expectedRecords, bigQueryRecords);
  }

  public Attributes mockBQReader() {
    return mockBQReader(bqHost(), bqPort());
  }

  public Attributes mockBQReader(int port) {
    return mockBQReader(bqHost(), port);
  }

  public Attributes mockBQReader(String host, int port) {
    GrpcBigQueryStorageReader reader =
        GrpcBigQueryStorageReader.create(BigQueryStorageSettings.create(host, port), system());
    return BigQueryStorageAttributes.reader(reader);
  }
}
